src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+        node_id_to_data(gateway_node_id, &target[HDR_GATEWAY_OFS]);
+        node_id_to_data(sequence, &target[HDR_TIMESTAMP_OFS]);
+        error_to_data(msg.mti, &target[HDR_LEN + MSG_MTI_OFS]);
+        node_id_to_data(msg.src.id, &target[HDR_LEN + MSG_SRC_OFS]);
+        if (has_dst)
+        {
+            node_id_to_data(msg.dst.id, &target[HDR_LEN + MSG_DST_OFS]);
+            memcpy(&target[HDR_LEN + MSG_ADR_PAYLOAD_OFS], msg.payload.data(),
+                msg.payload.size());
+        }
+        else
+        {
+            memcpy(&target[HDR_LEN + MSG_GLOBAL_PAYLOAD_OFS],
+                msg.payload.data(), msg.payload.size());
+        }
+    }
+
+    /// Guesses the length of a tcp message from looking at the prefix of the
+    /// payload.
+    ///
+    /// @param data is the prefix of the incoming stream.
+    /// @param len is the number of available bytes in this prefix.
+    /// @return -1 if there is not enough bytes in the prefix yet to know the
+    /// length. Otherwise the total number of bytes that constitute this
+    /// packet.
+    static int get_tcp_message_len(const void *data, size_t len)
+    {
+        if (len < HDR_SIZE_END)
+        {
+            return -1;
+        }
+        const uint8_t *src = static_cast<const uint8_t *>(data);
+        uint32_t sz = src[HDR_SIZE_OFS];
+        sz <<= 8;
+        sz |= src[HDR_SIZE_OFS + 1];
+        sz <<= 8;
+        sz |= src[HDR_SIZE_OFS + 2];
+        return sz + HDR_SIZE_END;
+    }
+
+    /// Parses a TCP message format (from binary payload) into a general
+    /// OpenLCB message.
+    /// @param src the rendered TCP message.
+    /// @param tgt the output generic message
+    /// @return false if the message is not well formatted or
+    /// it is not an OpenLCB message.
+    static bool parse_tcp_message(const string &src, GenMessage *tgt)
+    {
+        int expected_size = get_tcp_message_len(src.data(), src.size());
+        if (expected_size > (int)src.size() || expected_size < MIN_MESSAGE_SIZE)
+        {
+            LOG(WARNING, "Incomplete or incorrectly formatted TCP message.");
+            return false;
+        }
+        uint16_t flags = data_to_error(&src[0]);
+        if ((flags & FLAGS_OPENLCB_MSG) == 0)
+        {
+            return false;
+        }
+        HASSERT((flags & FLAGS_CHAINING) == 0);
+        tgt->flagsDst = 0;
+        tgt->flagsSrc = 0;
+        if (flags & FLAGS_FRAGMENT_NOT_FIRST)
+        {
+            tgt->set_flag_dst(GenMessage::DSTFLAG_NOT_FIRST_MESSAGE);
+        }
+        if (flags & FLAGS_FRAGMENT_NOT_LAST)
+        {
+            tgt->set_flag_dst(GenMessage::DSTFLAG_NOT_LAST_MESSAGE);
+        }
+        // Contents of the size field in the header.
+        uint32_t sz = (uint8_t)src[HDR_SIZE_OFS];
+        sz <<= 8;
+        sz |= (uint8_t)src[HDR_SIZE_OFS + 1];
+        sz <<= 8;
+        sz |= (uint8_t)src[HDR_SIZE_OFS + 2];
+        if ((sz + HDR_SIZE_OFS) > src.size())
```

Comment:
Is this sequence not a repeat of the get_tcp_message_len() operation?  Don't write the same code twice?

Reply:
actually it's worse. It's also buggy, and since we have already called get_tcp_message_len earlier in this function this code is not even needed.
Deleted.